edb/server/protocol/auth_ext/_static/webauthn-register.js

```diff
@@ -0,0 +1,188 @@
+import { decodeBase64Url, encodeBase64Url } from "./utils.js";
+
+document.addEventListener("DOMContentLoaded", () => {
+  const registerForm = document.getElementById("email-factor");
+
+  if (registerForm === null) {
+    return;
+  }
+
+  registerForm.addEventListener("submit", async (event) => {
+    if (event.submitter?.id !== "webauthn-signup") {
+      return;
+    }
+    event.preventDefault();
+
+    const formData = new FormData(/** @type {HTMLFormElement} */ registerForm);
+    const email = formData.get("email");
+    const provider = "builtin::local_webauthn";
+    const challenge = formData.get("challenge");
+    const redirectOnFailure = formData.get("redirect_on_failure");
+    const redirectTo = formData.get("redirect_to");
+    const verifyUrl = formData.get("verify_url");
+
+    if (redirectTo === null) {
+      throw new Error("Missing redirect_to parameter");
+    }
+
+    try {
+      const maybeCode = await register({
+        email,
+        provider,
+        challenge,
+        verifyUrl,
+      });
+
+      const redirectUrl = new URL(redirectTo);
+      redirectUrl.searchParams.append("isSignup", "true");
+      if (maybeCode !== null) {
+        redirectUrl.searchParams.append("code", maybeCode);
+      }
+
+      window.location.href = redirectUrl.href;
+    } catch (error) {
+      console.error("Failed to register WebAuthn credentials:", error);
+      const url = new URL(redirectOnFailure ?? redirectTo);
+      url.searchParams.append("error", error.message);
+      window.location.href = url.href;
+    }
+  });
+});
+
+const WEBAUTHN_OPTIONS_URL = new URL(
+  "../webauthn/register/options",
+  window.location
+);
+const WEBAUTHN_REGISTER_URL = new URL("../webauthn/register", window.location);
+
+/**
+ * Register a new WebAuthn credential for the given email address
+ * @param {Object} props - The properties for registration
+ * @param {string} props.email - Email address to register
+ * @param {string} props.provider - WebAuthn provider
+ * @param {string} props.challenge - PKCE challenge
+ * @param {string} props.verifyUrl - URL to verify email after registration
+ * @returns {Promise<string | null>} - The PKCE code or null if the application
+ *   requires email verification
+ */
+export async function register({ email, provider, challenge, verifyUrl }) {
+  // Check if WebAuthn is supported
+  if (!window.PublicKeyCredential) {
+    console.error("WebAuthn is not supported in this browser.");
```

Comment:
Maybe we could do this check on page load, before the user tries to register? But we can probably leave that for a later PR when we've decided on the ui design.

Reply:
Yeah, I agree that we should hide any WebAuthn affordances if we detect that you cannot use WebAuthn. I suppose in the case where an application has decided to do WebAuthn only, we need to show some kind of helpful empty state as well.